### PR TITLE
fix: parseNamespace for <= 10 byte namespace IDs

### PR DESCRIPTION
**Pull request by rootulp**
Closes https://github.com/celestiaorg/celestia-node/issues/2321

### CODE REVIEW

File: cmd/celestia/rpc_test.go
Comment (rootulp) [inline cmd/celestia/rpc_test.go:37]: I think there are other bugs lurking in `parseNamespace` but I wanted to help unblock @jcstein

File: cmd/celestia/rpc.go
Comment (rootulp) [inline cmd/celestia/rpc.go:444]: [question] does the RPC CLI let user's provide a namespace version? If not, is support for that being tracked somewhere? It's not urgent b/c there is only one namespace version (`0`) but I think we may want to expose it as an optional CLI flag b/c it impacts how namespaces are constructed here

Comment (Wondertan) [reply cmd/celestia/rpc.go:444]: @distractedm1nd

File: cmd/celestia/rpc.go
Comment (Wondertan) [inline cmd/celestia/rpc.go:445]: This check happens in the constructor so it's redundant

**Comment by Wondertan** [comment]
@rootulp, note that I am waiting for @distractedm1nd to return from sick leave and look deeper into this PR, as he is an author. I am not deep in the context of rpc cli

File: cmd/celestia/rpc.go
Comment (Wondertan) [reply cmd/celestia/rpc.go:445]: I assumed returning an error here was correct. Do we want to avoid throwing because we want to support full namespaces besides sub-namespaces(<=10)?

If so, I suggest adding logic that detects if a namespace is full or sub and returns instantiated namespace. 
Something like `share.DecodeNamespace([]byte) Namespace`, would enable anyone to take a namespace of any form and simply use it.

File: cmd/celestia/rpc.go
Comment (distractedm1nd) [reply cmd/celestia/rpc.go:444]: It's not tracked and not urgent until the next namespaceID versions are out. The CLI is here for providing the easiest way to interact/play with the RPC so I wouldn't add it until it is absolutely necessary

**Review by distractedm1nd — COMMENTED** [review]
Sent @rootulp a fix for the missing test case and added two additional ones, now the coverage is pretty good. Once applied i will approve

File: cmd/celestia/rpc.go
Comment (Wondertan) [reply cmd/celestia/rpc.go:445]: Ideally, we would be able to accept all these 3 cases, and something like Decode/Parse would help. Ideally, we would also move the Parse/Decode helper next to app's namespace. How does that sound to you?

Comment (rootulp) [reply cmd/celestia/rpc.go:445]: > Ideally, we would be able to accept all these 3 cases

Are these the cases you're considering:

version | ID | namespace | example
--- | --- | --- | ---
✅ | ✅ | computed based on other params | `celestia rpc blob Submit 0x42690c204d39600fddd3 --namespace-version 0`
❌ (defaults to `0`) | ✅  | computed based on other params | `celestia rpc blob Submit 0x42690c204d39600fddd3`
❌ | ❌ | ✅ | `celestia rpc blob Submit 0x0000000000000000000000000000000000000001010101010101010101`

If so, I think it's confusing that the param in most cases accepts a namespace ID but in the last case accepts a namespace.

> Ideally, we would also move the Parse/Decode helper next to app's namespace

I'm onboard with extracting a helper from celestia-node to celestia-app if that's what is suggested. And/or moving it to a share or namespace package inside this repo. Defer to ya'll on where `parseNamespace` should live.

**Review by distractedm1nd — APPROVED** [review]

File: cmd/celestia/rpc.go
Comment (walldiss) [inline cmd/celestia/rpc.go:451]: Looks like you meany `>` instead of `<`
```suggestion
	} else if len(nID) > appns.NamespaceSize {
		return nil, fmt.Errorf("passed namespace is too large")
```

**Comment by rootulp** [comment]
Synced w/ @distractedm1nd offline and `parseNamespace` is intended to only be used for the **namespace ID** portion of a namespace so a test case in this PR is incorrect.

I'll push a few commits to this PR that renames things like namespace to **namespace ID** where applicable but we should really do https://github.com/celestiaorg/celestia-node/issues/2301 ASAP.

**Comment by rootulp** [comment]
The scope of this PR has increased b/c it wasn't clear to me what the original intended behavior was. Would appreciate extra attention when reviewing b/c this couples a fix, refactor, and behavior change (left pad instead of right pad)

File: cmd/celestia/rpc.go
Comment (rootulp) [reply cmd/celestia/rpc.go]: Also we really need to clarify to users if they should be providing a 10 byte namespace ID, a partial namespace ID (e.g. 8 bytes), or a full namespace.

**Comment by Wondertan** [comment]
> Synced w/ @distractedm1nd offline and parseNamespace is intended to only be used for the namespace ID portion of a namespace so a test case in this PR is incorrect.

@distractedm1nd, why do we want to disable versioning here?

File: cmd/celestia/rpc.go
Comment (Wondertan) [reply cmd/celestia/rpc.go]: passing this to @distractedm1nd as you will have a call today

Comment (Wondertan) [reply cmd/celestia/rpc.go]: From our sync discussion today:
> `namespace.Parse/Decode([]byte) Namespace` Which accepts:
    * Ver + Padded ID
    * Padded ID
    * ID
    * Maybe Ver + ID

> The text encoding is out of the scope of this function, so parseNamespace will stay, but just use it.

**Review by Wondertan — DISMISSED** [review]
LGTM

File: cmd/celestia/rpc.go
Comment (distractedm1nd) [reply cmd/celestia/rpc.go:116]: The idea here is that it is actually only parsing the ID portion of the namespace: hence NamespaceID instead of Namespace.

Comment (Wondertan) [reply cmd/celestia/rpc.go:116]: Ah, right, forgot about that, then ok

**Review by Wondertan — COMMENTED** [review]
I still don't get whats the point of panicking on user input though

File: cmd/celestia/rpc_test.go
Comment (rootulp) [inline cmd/celestia/rpc_test.go:20]: [blocking] this claims to right pad but the implementation claims to left pad. I think left padding is desired behavior

File: cmd/celestia/rpc.go
Comment (rootulp) [inline cmd/celestia/rpc.go:437]: [blocking] does this actually left pad? I think it right pads b/c the first test case

File: cmd/celestia/rpc_test.go
Comment (distractedm1nd) [reply cmd/celestia/rpc_test.go:20]: It actually left pads, but the unit tests are failing! Didn't catch until now

File: cmd/celestia/rpc.go
Comment (distractedm1nd) [reply cmd/celestia/rpc.go:437]: thanks

File: cmd/celestia/rpc_test.go
Comment (distractedm1nd) [reply cmd/celestia/rpc_test.go]: fixed in [4a098ab](https://github.com/celestiaorg/celestia-node/pull/2325/commits/4a098abbd87ef34d5dbb7315a0d4cf52d55b379f)

**Review by Wondertan — APPROVED** [review]
Good catch @rootulp

**Review by renaynay — APPROVED** [review]

**Pull request merged**
